duprequests/middleware.py

```diff
@@ -1,41 +1,56 @@
-
-from uuid import uuid4
+from uuid import uuid4, uuid5, NAMESPACE_DNS
+from urllib.parse import urlencode
 
 from django.conf import settings
 from django.core.cache import caches
 from django.http.response import HttpResponseNotModified
 
+
 try:
     from django.utils.deprecation import MiddlewareMixin
-except ImportError:   # pragma: nocover
+except ImportError:  # pragma: nocover
     MiddlewareMixin = object
 
-CACHE_NAME = getattr(settings, 'DUPLICATED_REQUESTS_CACHE_NAME', 'default')
-CACHE_TIMEOUT = getattr(settings, 'DUPLICATED_REQUESTS_CACHE_TIMEOUT', 5)
-COOKIE_NAME = getattr(settings, 'DUPLICATED_REQUESTS_COOKIE_NAME',
-                      'dj-request-id')
-COOKIE_PREFIX = getattr(settings, 'DUPLICATED_REQUESTS_COOKIE_PREFIX',
-                        'request-id-')
+CACHE_NAME = getattr(settings, "DUPLICATED_REQUESTS_CACHE_NAME", "default")
+CACHE_TIMEOUT = getattr(settings, "DUPLICATED_REQUESTS_CACHE_TIMEOUT", 5)
+COOKIE_NAME = getattr(settings, "DUPLICATED_REQUESTS_COOKIE_NAME", "dj-request-id")
+COOKIE_PREFIX = getattr(settings, "DUPLICATED_REQUESTS_COOKIE_PREFIX", "request-id-")
 
 
 class DropDuplicatedRequests(MiddlewareMixin):
-    """Middleware that drops requests made in quick succession.
-
-    Uses Django's caching system to check/save each request."""
+    """
+    Middleware that drops requests made in quick succession.
+    Uses Django's caching system to check/save each request.
+    """
+
+    @staticmethod
+    def _get_cookie_name(request):
+        hash_value = uuid5(
+            NAMESPACE_DNS,
+            request.path
+            + request.method.lower()
+            + urlencode(request.GET)
+            + '--'
+            + urlencode(request.POST),
+        ).node
+        return f"{COOKIE_NAME}-{hash_value}"
 
     def process_request(self, request):
-        if not request.method.lower() in ('post', 'put', 'delete', 'patch'):
+        if not request.method.lower() in ("post", "put", "delete", "patch"):
             return
 
-        cache_key = request.COOKIES.get(COOKIE_NAME)
-        if not cache_key:
+        cookie_name = self._get_cookie_name(request)
+        cookie_value = request.COOKIES.get(cookie_name)
+        if not cookie_value:
             return
 
+        cache_key = cookie_name + cookie_value
         cache = caches[CACHE_NAME]
         if cache_key in cache:
             return HttpResponseNotModified()
         cache.set(cache_key, True, CACHE_TIMEOUT)
 
     def process_response(self, request, response):
-        response.set_cookie(COOKIE_NAME, COOKIE_PREFIX + uuid4().hex)
+        cookie_name = self._get_cookie_name(request)
+        response.set_cookie(cookie_name, COOKIE_PREFIX + uuid4().hex)
```

Comment:
Since we're setting a different cookie name for each request/payload we have a concern which is the cookie limit per user agent (browser and apps) which is referenced by RFC 6265. Each user agent has a minimum cookie limit of 50 cookies per domain, but for best practice we would keep this value minimum possible to minimize network bandwidth.
We should probably expire these cookies within few seconds and then we would not have this problem.

From django docs:

`HttpResponse.set_cookie(key, value='', max_age=None, expires=None, path='/', domain=None, secure=False, httponly=False, samesite=None)`

`max_age should` be a number of seconds, or None (default) if the cookie should last only as long as the client’s browser session. If expires is not specified, it will be calculated.

RFC 6265 link:
https://datatracker.ietf.org/doc/rfc6265/?include_text=1

Reply:
![image](https://user-images.githubusercontent.com/28119777/90635660-d6826000-e1ff-11ea-8593-b0adfac227b2.png)

Login as user result
Expires with session (3 months)